magic_items/routes.py

```diff
@@ -0,0 +1,304 @@
+from flask import Blueprint, render_template, g, flash, redirect, request
+from magic_items.services import (
+    MagicItemService,
+    ItemForm,
+    ItemFilterForm,
+    CollectionAddForm,
+    CollectionService,
+    ItemCollectionService,
+)
+from users.services import UserService
+
+magic_routes = Blueprint("magic_routes", __name__)
+
+
+def before_manage_magic_items(func):
+    def wrapper(*args, **kwargs):
+        if not g.user:
+            flash("Please login or signup to manage magic items", "warning")
+            return redirect("/login")
+        return func(*args, **kwargs)
+
+    wrapper.__name__ = func.__name__ + wrapper.__name__
+    return wrapper
+
+
+def before_manage_collections(func):
+    def wrapper(*args, **kwargs):
+        if not g.user:
+            flash("Please login or signup to manage collections", "warning")
+            return redirect("/login")
+        return func(*args, **kwargs)
+
+    wrapper.__name__ = func.__name__ + wrapper.__name__
+    return wrapper
```

Comment:
These two decorators are doing the same thing, the only change here is your error message? That's a sign that you can use a more generic message like "You don't have permission to perform this action".